c_api/tests/c_api_behavior_test.cpp

```diff
@@ -113,3 +113,453 @@ TEST_F(CapiTest, TestBehaviorGetRuntimeCapabilities) {
+      {
+        ti_allocate_memory(TI_NULL_HANDLE, &allocate_info);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+    }
+  };
+
+  inner(TI_ARCH_VULKAN);
+}
+
+TEST_F(CapiTest, TestBehaviorFreeMemory) {
+  auto inner = [](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
+      TiRuntime runtime = ti_create_runtime(arch);
+      TiMemoryAllocateInfo *allocate_info = new TiMemoryAllocateInfo;
+      allocate_info->size = 1024;
+      allocate_info->usage = TI_MEMORY_USAGE_STORAGE_BIT;
+      TiMemory memory = ti_allocate_memory(runtime, allocate_info);
+      ti_free_memory(runtime, memory);
+      CHECK_TAICHI_SUCCESS();
+
+      // runtime & allocate_info are both null
+      {
+        ti_free_memory(TI_NULL_HANDLE, nullptr);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+
+      // runtime is null and allocate_info is valid
+      {
+        TiMemoryAllocateInfo *allocate_info = new TiMemoryAllocateInfo;
+        allocate_info->size = 1024;
+        allocate_info->usage = TI_MEMORY_USAGE_STORAGE_BIT;
+        TiMemory memory = ti_allocate_memory(runtime, allocate_info);
+        ti_free_memory(TI_NULL_HANDLE, memory);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+
+      // runtime is not null and allocate_info is null
+      {
+        TiRuntime runtime = ti_create_runtime(arch);
+        ti_free_memory(runtime, nullptr);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+    }
+  };
+  inner(TI_ARCH_VULKAN);
+}
+
+TEST_F(CapiTest, TestBehaviorMapMemory) {
+  TiMemoryAllocateInfo allocate_info;
+  allocate_info.size = 1024;
+  allocate_info.usage = TI_MEMORY_USAGE_STORAGE_BIT;
+
+  auto inner = [&](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
+      TiRuntime runtime = ti_create_runtime(arch);
+      TiMemory memory = ti_allocate_memory(runtime, &allocate_info);
+      ti_map_memory(runtime, memory);
+      CHECK_TAICHI_SUCCESS();
+
+      // runtime & memory are both null
+      {
+        ti_map_memory(TI_NULL_HANDLE, TI_NULL_HANDLE);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+
+      // runtime is null, memory is valid
+      {
+        TiRuntime runtime = ti_create_runtime(arch);
+        TiMemory memory = ti_allocate_memory(runtime, &allocate_info);
+        ti_map_memory(TI_NULL_HANDLE, memory);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+
+      // runtime is valid, memory is null
+      {
+        TiRuntime runtime = ti_create_runtime(arch);
+        ti_map_memory(runtime, TI_NULL_HANDLE);
+        CHECK_TAICHI_ERROR_IS(TI_ERROR_ARGUMENT_NULL);
+      }
+    }
+  };
+  inner(TI_ARCH_VULKAN);
+}
+
+TEST_F(CapiTest, TestBehaviorUnmapMemory) {
+  TiMemoryAllocateInfo allocate_info;
+  allocate_info.size = 1024;
+  auto inner = [&](TiArch arch) {
+    if (ti::is_arch_available(arch)) {
+      TiRuntime runtime = ti_create_runtime(arch);
+      for (int i = 0; i < 4; ++i) {
```

Comment:
Why loop 4?

Reply:
Four kinds of different memory usage bit.